### PR TITLE
Add `allowUnorderedMigrations` option to Migrator

### LINKED ISSUE #697: Out of Order Migrations and Possible Solutions
Hi Kysely team :wave:

First of all, I want to thank you for the awesome library. I've never seen such thorough type completion in a database library like this—it's incredible.

I'm using Kysely on a project at work and we're having an issue with migrations. Specifically, running migrations "out of order" and the [`ensureMigrationsNotCorrupted`](https://github.com/kysely-org/kysely/blob/823b0fb31394e6841bc459cf602cff622a74d3bc/src/migration/migrator.ts#L470) check. There was some discussion in https://github.com/kysely-org/kysely/issues/472 about this, but I wanted to open a new issue and brainstorm some possible solutions.

### The Problem

We have multiple developers on our team. If two people create migrations asynchronously on separate branches, either those migrations have to be merged in the right order or we have to rebase and manually rename the "out of order" files before merging.

Example:
- Developer A creates migration `2023-09-13-120000-migration-A.ts`
- Developer B creates migration `2023-09-14-120000-migration-B.ts`
- Both developers make a PR at roughly the same time
- Developer B gets their PR approved, merged, and deployed first
- Developer A then get their PR approved and merged
- Developer A deploys to production and tries to run migrations

Currently, `ensureMigrationsNotCorrupted` would throw an error. But this is an extremely common scenario when developing on a team. 

As a work-around right now, I'm trying to create a GitHub Action that will detect new migration files when a PR is merged, update the timestamp in the filename, and commit the changes. But it's complicated, has to be set up for each new project, and the author of any migrations has to roll back their migrations on their dev branch before checking out the main branch again. 

We could also manually rebase and rename the files before merging, but that's tedius and prone to being forgotten.

### Possible Solutions

Instead of work-arounds, I would prefer to help fix this for all Kysely users. I imagine this will come up over and over again for teams trying to adopt Kysely.

Here are a couple of ideas I have. I wanted to see if you liked either of them before trying to put a PR together.

1. Give users a configurable option for verifying migration order: `verifyMigrationOrder: 'strict'` and `verifyMigrationOrder: 'permissive'`. This seems like the easiest method from a user's perspective. Strict by default, permissive when you need it.

2. Allow users to pass in a `MigrationNotCorruptedProvider`, similar to how you have `FileMigrationProvider`. This seems more prone to user error but would be more customizable.

Anyways, sorry for the long-winded explanation, I just wanted to make sure I laid everything out clearly. This project is awesome and I'd love to help make it easier for people to adopt and use. Thanks!

### PR DESCRIPTION
**Pull request by tracehelms**
### What is this?

This add an `allowUnorderedMigrations` option to [`MigratorProps`](https://github.com/kysely-org/kysely/blob/1538485c/src/migration/migrator.ts#L598). The options are `true` or `false`, with `false` being the default.

When true, it will allow new migrations to be run even if they are added alphabetically before ones that have already executed.

When false, it will give an error if your migration files are not alphabetically ordered in the same order that they were executed in the database. 

Closes #697

### Why?

If a team has multiple developers, two people can create migrations asynchronously on separate branches. With strict ordering, those migration files have to be merged in the right order (or renamed before merging) or else Kysely will throw an error when trying to migrate them.

Example:

    - Developer A creates migration `2023-09-13-migration-A.ts`
    - Developer B creates migration `2023-09-14-migration-B.ts`
    - Both developers make a PR at roughly the same time
    - Developer B merges their PR and deploys
    - Developer A merges their PR
    - Developer A deploys and tries to run migrations

Kysely will give an error for this with strict ordering. With permissive ordering, the unrun migration is executed.

### Pending migrations vs. executed migrations

This also adds the concept of pending (unexecuted) migrations to Migrator. When migrating up, Migrator will run pending migrations in order for `n` steps. When migrating down, Migrator will consult the database for executed migrations and travel down that list `n` steps. 

This doesn't change the current functionality for ensuring ordered migrations, but also allows out of order ones.

__Some examples:__ 
_(`*` denotes current migration version)_

New migration:
```
Migrations Already Ran:  0  1  2*
Migration Files:         0  1  2  3

Migrate up:              0  1  2  3*
```

Migrate down one:
```
Migrations Already Ran:  0  1  2*
Migration Files:         0  1  2  3

Migrate down:            0  1*
```

New out-of-order migration:
```
Migrations Already Ran:        0  1  2     4*
Migration Files:               0  1  2  3  4  5

Migrate up (strict):           error
Migrate up (permissive):       0  1  2  3  4*
Migrate up again (permissive): 0  1  2  3  4  5*
```

Migrate down with out-of-order migration:
```
Migrations Already Ran:    0  1  2     4*
Migration Files:           0  1  2  3  4  5

Migrate down (strict):     error
Migrate down (permissive): 0  1  2*
```

### CODE REVIEW

**Comment by igalklebanov** [comment]
Hey 👋 

Thanks! 💪 

I feel like `allowUnorderedMigrations` seems clearer.

**Comment by tracehelms** [comment]
> I feel like allowUnorderedMigrations seems clearer, or am I missing something?

That would definitely be clearer, good thinking. I'll update and post up the changes.

**Comment by tracehelms** [comment]
@igalklebanov Done, and I updated the docs / comments too 👍

**Comment by tracehelms** [comment]
Hi @igalklebanov and @koskimas, I was just wondering if you've gotten a chance to look at this and what you think. Let me know if you've got any questions / requests 😄 Thanks!

**Comment by danolife** [comment]
Yes can you please release this already 😭 
My production is broken because of unordered migrations, I don't want to update everything manually again! 
Please @igalklebano @koskimas 🙏

**Comment by tracehelms** [comment]
This branch was out of date with `master` so I rebased and pushed up the changes 👍

**Comment by blbrown13** [comment]
This would be super helpful for our team's workflow! 😍  Hopefully it gets approved and merged soon 🙏🏻

**Comment by blbrown13** [comment]
This would be super helpful for our team's workflow! 😍  Hopefully it gets approved and merged soon (cc @igalklebanov @koskimas ) 🙏🏻

**Comment by avocadojesus** [comment]
This would be really useful for my team as well, coordinating migrations is very tricky for us and this would solve all of our problems. Other than this little snag, Kysely has been absolutely incredible to use, thanks so much to the maintainers for writing such an incredible library!!!

**Comment by koskimas** [comment]
Now that I have a week off from work, I finally have some time for Kysely and could review this one. 

@danolife We're not keeping theses PRs waiting just to be mean or because we hate your teams. We simply haven't had time.

@tracehelms This seems excellent and will definitely get merged! Could you rebase this one more time?

File: src/migration/migrator.ts
Comment (koskimas) [inline src/migration/migrator.ts:513]: This can fail if two migrations have been run within the same millisecond. Either make sure each migration has a different timestamp or order by name after ordering by timestmp.

Comment (koskimas) [reply src/migration/migrator.ts:513]: Doesn't ordering by timestamp mean that if you first run migrations

```
1, 2, 4
```

and then

```
3
```

after that, migrating down will run the down migration of `3` 

producing

```
1, 2, 4
```

Is this really expected?

Comment (koskimas) [reply src/migration/migrator.ts:513]: Actually, secondary order by name probably covers all realistic cases 🤔

**Comment by tracehelms** [comment]
Hey @koskimas, thanks for reviewing! I'll get this rebased and patched up with your changes. 👍

File: src/migration/migrator.ts
Comment (tracehelms) [reply src/migration/migrator.ts]: Ok, I added a secondary sort by name. Thanks for the suggestion!

**Comment by tracehelms** [comment]
Ready for another look @koskimas 👍

**Review by koskimas — APPROVED** [review]
Perfect!

**Comment by tracehelms** [comment]
Awesome, thanks for reviewing and merging! 💯

**Comment by koskimas** [comment]
This is now released in 0.27.2

**Comment by loremru** [comment]
Actually `allowUnorderedMigrations` doesn't work

image
image
image

**Comment by igalklebanov** [comment]
@loremru this PR contains a unit test that proves it does. Thus makes me highly doubt your error haa anything to do with this feature.

**Comment by loremru** [comment]
> @loremru this PR contains a unit test that proves it does. Thus makes me highly doubt your error has anything to do with this feature.

@igalklebanov 
I think #resolveMigrations() should not sort my migrations if I set `allowUnorderedMigrations` to true. The mistake definitely here

image

**Comment by loremru** [comment]
@igalklebanov 
I think `#resolveMigrations` should not sort my migrations if I set `allowUnorderedMigrations ` to true.
image

**Comment by igalklebanov** [comment]
@loremru you're misunderstanding the feature. it is about not throwing errors if a new migration file is added that is in the middle of the sorted ran migrations.

e.g. 
PR 1 has 0003-a.js
PR 2 has 0003-b.js

if PR 2 gets merged first, running the migrations once PR 1 is merged would fail because:

[x] 0001
[x] 0002
[ ] 0003-a
[x] 0003-b

this feature makes it not fail.

**Pull request merged**
